### PR TITLE
Update parse_help

### PR DESCRIPTION
**Pull request by oharan2**
python-benedict is great for these parts in our code.

### CODE REVIEW

File: rosa/cli.py [hunk 1]
```diff
@@ -77,43 +78,31 @@ def check_flag_in_flags(command_list, flag_str):
...
     _user_command = shlex.split(command)
     command = ["rosa"]
     command.extend(_user_command)
-    json_output = {}
-    auto_answer_yes = {}
-    auto_update = {}
-    aws_region_flag = {}
-    for cmd in command[1:]:
+    var_list = {
+        "json_output": "ojson",
```
Comment (rnetser) [inline rosa/cli.py:86]:
```suggestion
        "json_output": "-ojson",
```

File: rosa/cli.py [hunk 2]
```diff
@@ -77,43 +78,31 @@ def check_flag_in_flags(command_list, flag_str):
...
+    _allowed_commands = benedict.benedict(allowed_commands or parse_help())
     _user_command = shlex.split(command)
     command = ["rosa"]
     command.extend(_user_command)
-    json_output = {}
-    auto_answer_yes = {}
-    auto_update = {}
-    aws_region_flag = {}
-    for cmd in command[1:]:
+    var_list = {
```
Comment (rnetser) [inline rosa/cli.py:85]: please use meaningful names for variables

File: rosa/cli.py [hunk 3]
```diff
@@ -77,43 +78,31 @@ def check_flag_in_flags(command_list, flag_str):
...
+        "json_output": "ojson",
+        "auto_answer_yes": "--yes",
+        "auto_mode": "--mode=auto",
+    }
+
+    key_path = []
```
Comment (rnetser) [inline rosa/cli.py:91]: please use meaningful names for variables

File: rosa/cli.py [hunk 4]
```diff
@@ -77,43 +78,31 @@ def check_flag_in_flags(command_list, flag_str):
...
-        auto_answer_yes = _allowed_commands.get(cmd, auto_answer_yes.get(cmd, {}))
-        add_auto_answer_yes = auto_answer_yes.get("auto_answer_yes") is True
-        if add_auto_answer_yes:
-            command.append("--yes")
-
-        auto_update = _allowed_commands.get(cmd, auto_update.get(cmd, {}))
-        add_auto_update = auto_update.get("auto_mode") is True
-        if add_auto_update:
-            command.append("--mode=auto")
+        for key, value in var_list.items():
```
Comment (rnetser) [inline rosa/cli.py:97]: please use meaningful names for variables

File: rosa/cli.py [hunk 5]
```diff
@@ -77,43 +78,31 @@ def check_flag_in_flags(command_list, flag_str):
...
-        aws_region_flag = _allowed_commands.get(cmd, aws_region_flag.get(cmd, {}))
-        add_aws_region_flag = json_output.get("region") is True
-        if add_aws_region_flag and aws_region:
+        if _allowed_commands[key_path].get("region", {}) and aws_region:
             command.append(f"--region={aws_region}")
 
-        if any(
-            [add_json_output, add_auto_answer_yes, add_auto_update, add_aws_region_flag]
-        ):
+        if any(atr in _allowed_commands[key_path] for atr in var_list.keys()):
```
Comment (rnetser) [inline rosa/cli.py:104]: please use meaningful names for variables (what does `atr` stand for?)

File: rosa/cli.py [hunk 6]
```diff
@@ -152,99 +141,50 @@ def get_available_flags(command):
...
+    commands_dict = benedict.benedict()
+    var_list = {
+        "json_output": "-o, --output",
+        "auto_answer_yes": "-y, --yes",
+        "auto_mode": "-m, --mode",
+        "region": "--region",
+    }
+
+    def _fill_path(key_path):
+        for key, value in var_list.items():
```
Comment (rnetser) [inline rosa/cli.py:153]: please use meaningful names for variables

File: rosa/cli.py [hunk 7]
```diff
@@ -152,99 +141,50 @@ def get_available_flags(command):
...
-    commands_dict = {}
+    commands_dict = benedict.benedict()
+    var_list = {
+        "json_output": "-o, --output",
+        "auto_answer_yes": "-y, --yes",
+        "auto_mode": "-m, --mode",
+        "region": "--region",
+    }
+
+    def _fill_path(key_path):
```
Comment (rnetser) [inline rosa/cli.py:152]: please use meaningful names for functions

Comment (rnetser) [inline rosa/cli.py:152]: `key_path` is shadowed

File: rosa/cli.py [hunk 8]
```diff
@@ -152,99 +141,50 @@ def get_available_flags(command):
...
-                flag_str=auto_mode_str,
-            )
-            commands_dict[top_command]["region"] = check_flag_in_flags(
-                command_list=[rosa_cmd, top_command],
-                flag_str=region_str,
-            )
+            _fill_path(key_path)
 
-    return commands_dict
+    return dict(commands_dict)
```
Comment (rnetser) [inline rosa/cli.py:187]: `build_command` converts the dict back to benedict; why is this needed?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 8 above)
Comment (oharan2) [reply rosa/cli.py:187]: For the case that `parse_help` is called separately as standalone function, not as part of `execute`.

File: rosa/cli.py [hunk 9]
```diff
@@ -157,99 +150,54 @@ def get_available_flags(command):
...
+        _commands = get_available_commands(command=["rosa"] + flag_search_path)
 
         if _commands:
             # If top command has sub command
             for command in _commands:
-                commands_dict[top_command][command] = {}
-                _commands = get_available_commands(
-                    command=[rosa_cmd, top_command, command]
-                )
+                flag_search_path = [top_command, command]
```
Comment (rnetser) [inline rosa/cli.py:180]:
```suggestion
                flag_search_path.append(command)
```

File: rosa/cli.py [hunk 10]
```diff
@@ -157,99 +150,54 @@ def get_available_flags(command):
...
-                        ] = check_flag_in_flags(
-                            command_list=[rosa_cmd, top_command, _command],
-                            flag_str=auto_mode_str,
-                        )
-                        commands_dict[top_command][command][_command][
-                            "region"
-                        ] = check_flag_in_flags(
-                            command_list=[rosa_cmd, top_command, _command],
-                            flag_str=region_str,
+                        flag_search_path = [top_command, command, _command]
```
Comment (rnetser) [inline rosa/cli.py:186]:
```suggestion
                        flag_search_path.append(_command)
```

File: rosa/cli.py (hunk 8 above)
Comment (rnetser) [reply rosa/cli.py:200]: benedict has extended functionality but also provides anything dict provides - why not return the object as is?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 9 above)
Comment (oharan2) [reply rosa/cli.py:180]: Append affect the functionality since it make `flag_search_path` to hold the last command from the last loop iteration.

Example:
Desired output:
'create': {
'account-roles': {'json_output': False, 'auto_answer_yes': True, 'auto_mode': True, 'region': False}, 
'admin': {'json_output': True, 'auto_answer_yes': True, 'auto_mode': False, 'region': True}
...}

`flag_search_path` was ['create','account-roles'] and THEN ['create','admin']

Output With append:
`flag_search_path` = ['create','account-roles','admin']

'create': {
'account-roles': {
'json_output': False, 'auto_answer_yes': True, 'auto_mode': True, 'region': `False, 'admin':` {'json_output': False, 'auto_answer_yes': True, 'auto_mode': True, 'region': `False, 'cluster': `{
..}}}}}}

File: rosa/cli.py (hunk 10 above)
Comment (oharan2) [reply rosa/cli.py:186]: Ditto

File: rosa/cli.py (hunk 9 above)
Comment (oharan2) [reply rosa/cli.py:180]: @rnetser

Comment (rnetser) [reply rosa/cli.py:180]: so please rename arg, do not overwrite the top one with something else

File: rosa/cli.py (hunk 10 above)
Comment (rnetser) [reply rosa/cli.py:186]: ditto

**Review by rnetser — CHANGES_REQUESTED** [review]

**Comment by rnetser** [comment]
@oharan2 https://github.com/RedHatQE/rosa-python-client/pull/46 was merged; this PR can be closed.

File: rosa/cli.py [hunk 11]
```diff
@@ -140,98 +140,33 @@ def get_available_flags(command):
...
+        sub_commands = get_available_commands(command=["rosa"] + commands_search_path)
+        if not sub_commands:
+            return _fill_commands_dict_with_support_flags(
+                flag_key_path=commands_search_path
             )
+        for sub_command in sub_commands:
+            commands_dict[commands_search_path, sub_command] = {}
+            _build_command_tree(commands_search_path + [sub_command])
+
+    _build_command_tree([])
```
Comment (myakove) [inline rosa/cli.py:169]: `([])` this doesn't make sense, please set `commands_search_path=None` in `_build_command_tree`

File: rosa/cli.py [hunk 12]
```diff
@@ -140,98 +140,33 @@ def get_available_flags(command):
...
+
+    def _build_command_tree(commands_search_path):
+        sub_commands = get_available_commands(command=["rosa"] + commands_search_path)
+        if not sub_commands:
+            return _fill_commands_dict_with_support_flags(
+                flag_key_path=commands_search_path
             )
+        for sub_command in sub_commands:
+            commands_dict[commands_search_path, sub_command] = {}
+            _build_command_tree(commands_search_path + [sub_command])
```
Comment (myakove) [inline rosa/cli.py:167]: this `commands_search_path + [sub_command]` same as `[commands_search_path, sub_command]` no?

File: rosa/cli.py [hunk 13]
```diff
@@ -140,98 +140,33 @@ def get_available_flags(command):
...
-            )
-            commands_dict[top_command]["auto_mode"] = check_flag_in_flags(
-                command_list=[rosa_cmd, top_command],
-                flag_str=auto_mode_str,
-            )
-            commands_dict[top_command]["region"] = check_flag_in_flags(
-                command_list=[rosa_cmd, top_command],
-                flag_str=region_str,
+
+    def _build_command_tree(commands_search_path):
```
Comment (myakove) [inline rosa/cli.py:159]: `_build_command_tree` and `_fill_commands_dict_with_support_flags` should get `commands_dict` arg

**Review by myakove — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 12 above)
Comment (oharan2) [reply rosa/cli.py]: No, for example:
sub_command: 'account-roles'
commands_search_path = ['delete']

[commands_search_path, sub_command] = `[['delete'], 'account-roles']`
commands_search_path + [sub_command] = `['delete', 'account-roles']` **- desired result**

File: rosa/cli.py [hunk 14]
```diff
@@ -155,98 +155,36 @@ def get_available_flags(command):
...
+        if not commands_search_path:
+            commands_search_path = []
+        sub_commands = get_available_commands(command=["rosa"] + commands_search_path)
+        if not sub_commands:
+            return _fill_commands_dict_with_support_flags(
+                commands_dict=commands_dict, flag_key_path=commands_search_path
+            )
+        for sub_command in sub_commands:
+            commands_dict[commands_search_path, sub_command] = {}
+            _build_command_tree(commands_dict, commands_search_path + [sub_command])
```
Comment (rnetser) [inline rosa/cli.py:183]: please use arg=val

**Review by rnetser — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 14 above)
Comment (oharan2) [reply rosa/cli.py]: Done

File: rosa/cli.py [hunk 15]
```diff
@@ -160,98 +160,40 @@ def get_available_flags(command):
...
-                commands_dict[top_command][command]["auto_mode"] = check_flag_in_flags(
-                    command_list=[rosa_cmd, top_command, command],
-                    flag_str=auto_mode_str,
-                )
-                commands_dict[top_command][command]["region"] = check_flag_in_flags(
-                    command_list=[rosa_cmd, top_command, command],
-                    flag_str=region_str,
-                )
+    def _build_command_tree(commands_dict, commands_search_path=None):
+        _commands_dict = commands_dict
```
Comment (myakove) [inline rosa/cli.py:179]: `_commands_dict = commands_dict` why you assign it?
It will be a reference to the same dict

**Review by myakove — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 15 above)
Comment (oharan2) [reply rosa/cli.py:179]: Just because I thought about avoiding calling with `commands_dict=commands_dict` in line [190](https://github.com/RedHatQE/rosa-python-client/pull/34/files#diff-d727c959dd93ddbb1b1b2d0cd8124ec5d587ebc047afe9f117e2d7e7858fbd23R190)
What do you think?

Comment (myakove) [reply rosa/cli.py:179]: This assignment is useless and unnecessary, it's just point to the same object.

**Review by myakove — CHANGES_REQUESTED** [review]

File: rosa/cli.py (hunk 15 above)
Comment (oharan2) [reply rosa/cli.py]: Done

**Review by rnetser — APPROVED** [review]
